roachtest: scrub/all-checks/tpcc-1000 failed

SHA: https://github.com/cockroachdb/cockroach/commits/859214b81838a4ba33048b81497442ce5774baa7

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc-1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1053297&tab=buildLog

```
The test failed on master:
	test.go:630,schemachange.go:329,scrub.go:75,cluster.go:1465,errgroup.go:57: dial tcp 35.232.175.103:26257: connect: connection refused
	test.go:630,cluster.go:1486,tpcc.go:120,scrub.go:71: Goexit() was called

```

SHA: https://github.com/cockroachdb/cockroach/commits/3c80c5e06c4aa5ed25e1cc02b78037f6ec121939

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc-1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1072755&tab=buildLog

```
	test.go:703,cluster.go:1137,tpcc.go:110,cluster.go:1463,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1072755-scrub-all-checks-tpcc-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		           0.7      0.0      0.0      0.0      0.0 delivery
		    6m7s      798            0.0            3.8      0.0      0.0      0.0      0.0 newOrder
		    6m7s      798            0.0            0.7      0.0      0.0      0.0      0.0 orderStatus
		    6m7s      798            0.0            1.1      0.0      0.0      0.0      0.0 payment
		    6m7s      798            0.0            0.6      0.0      0.0      0.0      0.0 stockLevel
		E181229 14:09:38.875391 1 workload/cli/run.go:402  error in payment: ERROR: context deadline exceeded (SQLSTATE XX000)
		    6m8s      804            0.0            0.7      0.0      0.0      0.0      0.0 delivery
		    6m8s      804            0.0            3.8      0.0      0.0      0.0      0.0 newOrder
		    6m8s      804            0.0            0.7      0.0      0.0      0.0      0.0 orderStatus
		    6m8s      804            0.0            1.1      0.0      0.0      0.0      0.0 payment
		    6m8s      804            0.0            0.6      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	test.go:703,cluster.go:1484,tpcc.go:120,scrub.go:56: Goexit() was called

```

SHA: https://github.com/cockroachdb/cockroach/commits/b024b461265a7ca3cc1d156fef459818d127b065

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc-1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1074018&tab=buildLog

```
	test.go:703,cluster.go:1137,tpcc.go:110,cluster.go:1463,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1074018-scrub-all-checks-tpcc-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		 1232            0.0            0.2      0.0      0.0      0.0      0.0 orderStatus
		   7m35s     1232            0.0            0.2      0.0      0.0      0.0      0.0 payment
		   7m35s     1232            0.0            0.2      0.0      0.0      0.0      0.0 stockLevel
		E181231 14:46:42.138647 1 workload/cli/run.go:402  error in newOrder: ERROR: TransactionStatusError: transaction deadline exceeded (REASON_UNKNOWN) (SQLSTATE XX000)
		   7m36s     1234            0.0            0.2      0.0      0.0      0.0      0.0 delivery
		   7m36s     1234            0.0            0.7      0.0      0.0      0.0      0.0 newOrder
		   7m36s     1234            0.0            0.2      0.0      0.0      0.0      0.0 orderStatus
		   7m36s     1234            0.0            0.2      0.0      0.0      0.0      0.0 payment
		   7m36s     1234            0.0            0.2      0.0      0.0      0.0      0.0 stockLevel
		E181231 14:46:43.376830 1 workload/cli/run.go:402  error in payment: ERROR: context deadline exceeded (SQLSTATE XX000)
		: signal: killed
	test.go:703,cluster.go:1484,tpcc.go:120,scrub.go:56: Goexit() was called

```

SHA: https://github.com/cockroachdb/cockroach/commits/f3cf71db94327abb6a164ad67383c35a696ec7d8

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc-1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1077687&tab=buildLog

```
	test.go:696,cluster.go:1137,tpcc.go:110,cluster.go:1463,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1077687-scrub-all-checks-tpcc-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		     0.0            2.3      0.0      0.0      0.0      0.0 newOrder
		1h42m44s    20419            0.0            0.3      0.0      0.0      0.0      0.0 orderStatus
		1h42m44s    20419            0.0            1.3      0.0      0.0      0.0      0.0 payment
		1h42m44s    20419            0.0            0.3      0.0      0.0      0.0      0.0 stockLevel
		_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		1h42m45s    20419            0.0            0.4      0.0      0.0      0.0      0.0 delivery
		1h42m45s    20419            0.0            2.3      0.0      0.0      0.0      0.0 newOrder
		1h42m45s    20419            0.0            0.3      0.0      0.0      0.0      0.0 orderStatus
		1h42m45s    20419            0.0            1.3      0.0      0.0      0.0      0.0 payment
		1h42m45s    20419            0.0            0.3      0.0      0.0      0.0      0.0 stockLevel
		E190103 16:40:11.552241 1 workload/cli/run.go:402  error in payment: ERROR: context deadline exceeded (SQLSTATE XX000)
		: signal: killed
	test.go:696,cluster.go:1484,tpcc.go:120,scrub.go:58: Goexit() was called

```

SHA: https://github.com/cockroachdb/cockroach/commits/8f446e3a82c6d10965fa86a268ec96c94ac093ec

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc-1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1083973&tab=buildLog

```
	test.go:696,cluster.go:1164,tpcc.go:110,cluster.go:1490,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1083973-scrub-all-checks-tpcc-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		79.2 103079.2 newOrder
		1h50m40s    22995            1.0            0.3     32.5     32.5     32.5     32.5 orderStatus
		1h50m40s    22995            0.0            1.0      0.0      0.0      0.0      0.0 payment
		1h50m40s    22995            2.0            0.3    121.6    167.8    167.8    167.8 stockLevel
		E190108 15:18:53.374079 1 workload/cli/run.go:402  error in newOrder: ERROR: TransactionStatusError: transaction deadline exceeded (REASON_UNKNOWN) (SQLSTATE XX000)
		_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)
		1h50m41s    22998            0.0            0.3      0.0      0.0      0.0      0.0 delivery
		1h50m41s    22998           13.0            1.9  23622.3 103079.2 103079.2 103079.2 newOrder
		1h50m41s    22998            0.0            0.3      0.0      0.0      0.0      0.0 orderStatus
		1h50m41s    22998            0.0            1.0      0.0      0.0      0.0      0.0 payment
		1h50m41s    22998            0.0            0.3      0.0      0.0      0.0      0.0 stockLevel
		: signal: killed
	test.go:696,cluster.go:1511,tpcc.go:120,scrub.go:58: Goexit() was called

```

SHA: https://github.com/cockroachdb/cockroach/commits/395d842feb97c5bd8cad2b32b71a5156c03061eb

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scrub/all-checks/tpcc-1000 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1115923&tab=buildLog

```
The test failed on master:
	test.go:743,cluster.go:1226,tpcc.go:118,cluster.go:1564,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1115923-scrub-all-checks-tpcc-1000:5 -- ./workload run tpcc --warehouses=1000 --histograms=logs/stats.json --wait=false --tolerate-errors --ramp=5m0s --duration=2h0m0s {pgurl:1-4} returned:
		stderr:
		
		stdout:
		2 103079.2 newOrder
		   4m14s    58207            0.0            8.1      0.0      0.0      0.0      0.0 orderStatus
		   4m14s    58207            7.0           64.7  10200.5 103079.2 103079.2 103079.2 payment
		   4m14s    58207            1.0            8.1   1744.8   1744.8   1744.8   1744.8 stockLevel
		E190130 15:02:26.667367 1 workload/cli/run.go:402  error in newOrder: dial tcp 10.128.0.25:26257: connect: connection refused
		   4m15s   102271            3.0            7.8  12348.0  32212.3  32212.3  32212.3 delivery
		   4m15s   102271           95.8           70.5  12348.0 103079.2 103079.2 103079.2 newOrder
		   4m15s   102271            4.9            8.1    469.8   2080.4   2080.4   2080.4 orderStatus
		   4m15s   102271           29.6           64.5  11811.2 103079.2 103079.2 103079.2 payment
		   4m15s   102271            4.9            8.1   2684.4  10737.4  10737.4  10737.4 stockLevel
		E190130 15:02:27.667487 1 workload/cli/run.go:402  error in newOrder: dial tcp 10.128.0.25:26257: connect: connection refused
		: signal: killed
	test.go:743,cluster.go:1302,cluster.go:1321,cluster.go:1425,scrub.go:72,cluster.go:1564,errgroup.go:57: context canceled
	test.go:743,cluster.go:1585,tpcc.go:128,scrub.go:58: unexpected node event: 3: dead

```

Previous failure was
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x190242d]

goroutine 719576 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0001c5830, 0x3898dc0, 0xc00f811050)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:183 +0x11f
panic(0x2d24dc0, 0x546ff60)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/storage.(*truncateDecision).raftSnapshotsForIndex(0xc0060625f0, 0x0, 0xc0054edb00)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/raft_log_queue.go:226 +0x3d
```

Fixed by https://github.com/cockroachdb/cockroach/pull/34399.

--- a/pkg/storage/client_replica_test.go
+++ b/pkg/storage/client_replica_test.go
@@ -1841,6 +1841,7 @@ func TestLeaseTransferInSnapshotUpdatesTimestampCache(t *testing.T) {
 	mtc := &multiTestContext{storeConfig: &sc}
 	defer mtc.Stop()
 	mtc.Start(t, 3)
+	store2 := mtc.Store(2)
 
 	keyA := roachpb.Key("a")
 	keyB := roachpb.Key("b")
@@ -1905,7 +1906,11 @@ func TestLeaseTransferInSnapshotUpdatesTimestampCache(t *testing.T) {
 	// another write and truncate the Raft log on the two connected nodes. This
 	// ensures that that when node 2 comes back up it will require a snapshot
 	// from Raft.
-	mtc.transport.GetCircuitBreaker(mtc.idents[2].NodeID).Break()
+	mtc.transport.Listen(store2.Ident.StoreID, &unreliableRaftHandler{
+		rangeID:            rangeID,
+		RaftMessageHandler: store2,
+	})
+
 	if _, pErr := client.SendWrapped(ctx, mtc.stores[0].TestSender(), incC); pErr != nil {
 		t.Fatal(pErr)
 	}
@@ -1927,14 +1932,14 @@ func TestLeaseTransferInSnapshotUpdatesTimestampCache(t *testing.T) {
 	// Finally, transfer the lease to node 2 while it is still unavailable and
 	// behind. We try to avoid this case when picking new leaseholders in practice,
 	// but we're never 100% successful.
-	if err := repl0.AdminTransferLease(ctx, mtc.idents[2].StoreID); err != nil {
+	if err := repl0.AdminTransferLease(ctx, store2.Ident.StoreID); err != nil {
 		t.Fatal(err)
 	}
 
 	// Remove the partition. A snapshot to node 2 should follow. This snapshot
 	// will inform node 2 that it is the new leaseholder for the range. Node 2
 	// should act accordingly and update its internal state to reflect this.
-	mtc.transport.GetCircuitBreaker(mtc.idents[2].NodeID).Reset()
+	mtc.transport.Listen(store2.Ident.StoreID, store2)
 	mtc.waitForValues(keyC, []int64{4, 4, 4})
 
 	// Perform a write on the new leaseholder underneath the previously served

--- a/pkg/storage/client_replica_test.go
+++ b/pkg/storage/client_replica_test.go
@@ -1822,3 +1822,132 @@ func TestClearRange(t *testing.T) {
 	verifyKeysWithPrefix(keys.LocalStoreSuggestedCompactionsMin,
 		[]roachpb.Key{keys.StoreSuggestedCompactionKey(lg1, lg3)})
 }
+
+// TestLeaseTransferInSnapshotUpdatesTimestampCache prevents a regression of
+// #34025. A Replica is targeted for a lease transfer target when it needs a
+// Raft snapshot to catch up. Normally we try to prevent this case, but it is
+// possible and hard to prevent entirely. The Replica will only learn that it is
+// the new leaseholder when it applies the snapshot. When doing so, it should
+// make sure to apply the lease-related side-effects to its in-memory state.
+func TestLeaseTransferInSnapshotUpdatesTimestampCache(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	ctx := context.Background()
+	sc := storage.TestStoreConfig(nil)
+	// We'll control replication by hand.
+	sc.TestingKnobs.DisableReplicateQueue = true
+	// Avoid fighting with the merge queue while trying to reproduce this race.
+	sc.TestingKnobs.DisableMergeQueue = true
+	mtc := &multiTestContext{storeConfig: &sc}
+	defer mtc.Stop()
+	mtc.Start(t, 3)
+
+	keyA := roachpb.Key("a")
+	keyB := roachpb.Key("b")
+	keyC := roachpb.Key("c")
+
+	// First, do a couple of writes; we'll use these to determine when
+	// the dust has settled.
+	incA := incrementArgs(keyA, 1)
+	if _, pErr := client.SendWrapped(ctx, mtc.stores[0].TestSender(), incA); pErr != nil {
+		t.Fatal(pErr)
+	}
+	incC := incrementArgs(keyC, 2)
+	if _, pErr := client.SendWrapped(ctx, mtc.stores[0].TestSender(), incC); pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	// Split the system range from the rest of the keyspace.
+	splitArgs := adminSplitArgs(keys.SystemMax)
+	if _, pErr := client.SendWrapped(ctx, mtc.stores[0].TestSender(), splitArgs); pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	// Get the range's ID.
+	repl0 := mtc.stores[0].LookupReplica(roachpb.RKey(keyA))
+	rangeID := repl0.RangeID
+
+	// Replicate the range onto nodes 1 and 2.
+	// Wait for all replicas to be caught up.
+	mtc.replicateRange(rangeID, 1, 2)
+	mtc.waitForValues(keyA, []int64{1, 1, 1})
+	mtc.waitForValues(keyC, []int64{2, 2, 2})
+
+	// Create a transaction that will try to write "under" a served read.
+	// The read will have been served by the original leaseholder (node 0)
+	// and the write will be attempted on the new leaseholder (node 2).
+	// It should not succeed because it should run into the timestamp cache.
+	db := mtc.dbs[0]
+	txnOld := client.NewTxn(ctx, db, 0 /* gatewayNodeID */, client.RootTxn)
+
+	// Perform a write with txnOld so that its timestamp gets set and so
+	// that it writes its transaction record.
+	if err := txnOld.EagerRecord(); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := txnOld.Inc(ctx, keyB, 3); err != nil {
+		t.Fatal(err)
+	}
+
+	// Read keyC with txnOld, which is updated below. This prevents the
+	// transaction from refreshing when it hits the serializable error.
+	if _, err := txnOld.Get(ctx, keyC); err != nil {
+		t.Fatal(err)
+	}
+
+	// Another client comes along at a higher timestamp and reads. We should
+	// never be able to write under this time or we would be rewriting history.
+	if _, err := db.Get(ctx, keyA); err != nil {
+		t.Fatal(err)
+	}
+
+	// Partition node 2 from the rest of its range. Once partitioned, perform
+	// another write and truncate the Raft log on the two connected nodes. This
+	// ensures that that when node 2 comes back up it will require a snapshot
+	// from Raft.
+	mtc.transport.GetCircuitBreaker(mtc.idents[2].NodeID).Break()
+	if _, pErr := client.SendWrapped(ctx, mtc.stores[0].TestSender(), incC); pErr != nil {
+		t.Fatal(pErr)
+	}
+	mtc.waitForValues(keyC, []int64{4, 4, 2})
+
+	// Truncate the log at index+1 (log entries < N are removed, so this
+	// includes the increment). This necessitates a snapshot when the
+	// partitioned replica rejoins the rest of the range.
+	index, err := repl0.GetLastIndex()
+	if err != nil {
+		t.Fatal(err)
+	}
+	truncArgs := truncateLogArgs(index+1, rangeID)
+	truncArgs.Key = keyA
+	if _, err := client.SendWrapped(ctx, mtc.stores[0].TestSender(), truncArgs); err != nil {
+		t.Fatal(err)
+	}
+
+	// Finally, transfer the lease to node 2 while it is still unavailable and
+	// behind. We try to avoid this case when picking new leaseholders in practice,
+	// but we're never 100% successful.
+	if err := repl0.AdminTransferLease(ctx, mtc.idents[2].StoreID); err != nil {
+		t.Fatal(err)
+	}
+
+	// Remove the partition. A snapshot to node 2 should follow. This snapshot
+	// will inform node 2 that it is the new leaseholder for the range. Node 2
+	// should act accordingly and update its internal state to reflect this.
+	mtc.transport.GetCircuitBreaker(mtc.idents[2].NodeID).Reset()
+	mtc.waitForValues(keyC, []int64{4, 4, 4})
+
+	// Perform a write on the new leaseholder underneath the previously served
+	// read. This write should hit the timestamp cache and flag the txn for a
+	// restart when we try to commit it below. With the bug in #34025, the new
+	// leaseholder who heard about the lease transfer from a snapshot had an
+	// empty timestamp cache and would simply let us write under the previous
+	// read.
+	if _, err := txnOld.Inc(ctx, keyA, 4); err != nil {
+		t.Fatal(err)
+	}
+	const exp = `TransactionRetryError: retry txn \(RETRY_SERIALIZABLE\)`
+	if err := txnOld.Commit(ctx); !testutils.IsError(err, exp) {
+		t.Fatalf("expected retry error, got: %v; did we write under a read?", err)
+	}
+}

--- a/pkg/storage/replica_proposal.go
+++ b/pkg/storage/replica_proposal.go
@@ -190,11 +190,22 @@ func (r *Replica) computeChecksumPostApply(ctx context.Context, cc storagepb.Com
 	}
 }
 
-// leasePostApply is called when a RequestLease or TransferLease
-// request is executed for a range.
-func (r *Replica) leasePostApply(ctx context.Context, newLease roachpb.Lease) {
+// leasePostApply updates the Replica's internal state to reflect the
+// application of a new Range lease. The method is idempotent, so it can be
+// called repeatedly for the same lease safely. However, the method will panic
+// if passed a lease with a lower sequence number than the current lease. By
+// default, the method will also panic if passed a lease that indicates a
+// forward sequence number jump (i.e. a skipped lease). This behavior can
+// be disabled by passing permitJump as true.
+func (r *Replica) leasePostApply(ctx context.Context, newLease roachpb.Lease, permitJump bool) {
 	r.mu.Lock()
 	replicaID := r.mu.replicaID
+	// Pull out the last lease known to this Replica. It's possible that this is
+	// not actually the last lease in the Range's lease sequence because the
+	// Replica may have missed the application of a lease between prevLease and
+	// newLease. However, this should only be possible if a snapshot includes a
+	// lease update. All other forms of lease updates should be continuous
+	// without jumps (see permitJump).
 	prevLease := *r.mu.state.Lease
 	r.mu.Unlock()
 
@@ -267,9 +278,7 @@ func (r *Replica) leasePostApply(ctx context.Context, newLease roachpb.Lease) {
 			}
 		case s2 == s1+1:
 			// Lease sequence incremented by 1. Expected case.
-		case s2 > s1+1:
-			// Snapshots will never call leasePostApply, so we always expect
-			// leases to increment one at a time here.
+		case s2 > s1+1 && !permitJump:
 			log.Fatalf(ctx, "lease sequence jump, prevLease=%s, newLease=%s",
 				log.Safe(prevLease), log.Safe(newLease))
 		}
@@ -634,7 +643,7 @@ func (r *Replica) handleReplicatedEvalResult(
 		}
 
 		if newLease := rResult.State.Lease; newLease != nil {
-			r.leasePostApply(ctx, *newLease)
+			r.leasePostApply(ctx, *newLease, false /* permitJump */)
 			rResult.State.Lease = nil
 		}
 

--- a/pkg/storage/replica_raftstorage.go
+++ b/pkg/storage/replica_raftstorage.go
@@ -952,6 +952,12 @@ func (r *Replica) applySnapshot(
 	}
 	r.store.mu.Unlock()
 
+	// Invoke the leasePostApply method to ensure we properly initialize the
+	// replica according to whether it holds the lease. We allow jumps in the
+	// lease sequence because there may be multiple lease changes accounted for
+	// in the snapshot.
+	r.leasePostApply(ctx, *s.Lease, true /* permitJump */)
+
 	r.mu.Lock()
 	// We set the persisted last index to the last applied index. This is
 	// not a correctness issue, but means that we may have just transferred
@@ -967,7 +973,8 @@ func (r *Replica) applySnapshot(
 	r.store.metrics.subtractMVCCStats(*r.mu.state.Stats)
 	r.store.metrics.addMVCCStats(*s.Stats)
 	// Update the rest of the Raft state. Changes to r.mu.state.Desc must be
-	// managed by r.setDesc, but we called that above, so now it's safe to
+	// managed by r.setDesc and changes to r.mu.state.Lease must be handled
+	// by r.leasePostApply, but we called those above, so now it's safe to
 	// wholesale replace r.mu.state.
 	r.mu.state = s
 	r.assertStateLocked(ctx, r.store.Engine())

--- a/pkg/storage/replica_rangefeed_test.go
+++ b/pkg/storage/replica_rangefeed_test.go
@@ -474,9 +474,9 @@ func TestReplicaRangefeedRetryErrors(t *testing.T) {
 		assertRangefeedRetryErr(t, pErrRight, roachpb.RangeFeedRetryError_REASON_RANGE_MERGED)
 	})
 	t.Run(roachpb.RangeFeedRetryError_REASON_RAFT_SNAPSHOT.String(), func(t *testing.T) {
-		const partitionStore = 2
 		mtc, rangeID := setup(t)
 		defer mtc.Stop()
+		partitionStore := mtc.Store(2)
 
 		mtc.stores[0].SetReplicaGCQueueActive(false)
 		mtc.stores[1].SetReplicaGCQueueActive(false)
@@ -493,7 +493,7 @@ func TestReplicaRangefeedRetryErrors(t *testing.T) {
 				Span: roachpb.Span{Key: roachpb.Key("a"), EndKey: roachpb.Key("z")},
 			}
 
-			pErr := mtc.Store(partitionStore).RangeFeed(ctx, &req, stream)
+			pErr := partitionStore.RangeFeed(ctx, &req, stream)
 			streamErrC <- pErr
 		}()
 
@@ -501,7 +501,10 @@ func TestReplicaRangefeedRetryErrors(t *testing.T) {
 		waitForInitialCheckpoint(t, stream, streamErrC)
 
 		// Partition the replica from the rest of its range.
-		mtc.transport.GetCircuitBreaker(mtc.idents[partitionStore].NodeID).Break()
+		mtc.transport.Listen(partitionStore.Ident.StoreID, &unreliableRaftHandler{
+			rangeID:            rangeID,
+			RaftMessageHandler: partitionStore,
+		})
 
 		// Perform a write on the range.
 		pArgs := putArgs(roachpb.Key("c"), []byte("val2"))
@@ -529,7 +532,7 @@ func TestReplicaRangefeedRetryErrors(t *testing.T) {
 		}
 
 		// Remove the partition. Snapshot should follow.
-		mtc.transport.GetCircuitBreaker(mtc.idents[partitionStore].NodeID).Reset()
+		mtc.transport.Listen(partitionStore.Ident.StoreID, partitionStore)
 
 		// Check the error.
 		pErr := <-streamErrC

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -2312,7 +2312,7 @@ func splitPostApply(
 	// Invoke the leasePostApply method to ensure we properly initialize
 	// the replica according to whether it holds the lease. This enables
 	// the txnWaitQueue.
-	rightRng.leasePostApply(ctx, rightLease)
+	rightRng.leasePostApply(ctx, rightLease, false /* permitJump */)
 
 	// Add the RHS replica to the store. This step atomically updates
 	// the EndKey of the LHS replica and also adds the RHS replica

